src/blocks/upower.rs

```diff
@@ -0,0 +1,306 @@
+        let m = Self::get_property(c, device_path, "Type");
+
+        match m {
+            Err(_) => false,
+            Ok(msg) => {
+                let type_battery = DeviceType::Battery.to_u32();
+                let type_: Option<Variant<u32>> = msg.get1();
+
+                match type_ {
+                    Some(Variant(m)) if m == type_battery => true,
+                    _ => false,
+                }
+            },
+        }
+    }
+
+    pub fn percentage(&self, c: &Connection) -> Result<Option<u8>> {
+        let m = try!(Self::get_property(c, &self.get_device_path(), "IsPresent"));
+        let is_present: Variant<bool> = m.get1().block_error(
+            "upower", "Failed to read property"
+        )?;
+
+        if !is_present.0 {
+            return Ok(None)
+        }
+
+        let m = try!(Self::get_property(c, &self.get_device_path(), "Percentage"));
+        let percentage: Variant<f64> = m.get1().block_error(
+            "upower", "Failed to read property"
+        )?;
+
+        Ok(Some(percentage.0 as u8))
+    }
+
+    pub fn state(&self, c: &Connection) -> Result<BatteryState> {
+        let m = try!(Self::get_property(c, &self.get_device_path(), "State"));
+
+        let state: Variant<u32> = m.get1().block_error(
+            "upower", "Failed to read property"
+        )?;
+
+        Ok(BatteryState::from(state.0))
+    }
+}
+
+pub struct Upower {
+    id: String,
+    output: TextWidget,
+    dbus_conn: Connection,
+    battery: Battery,
+}
+
+#[derive(Deserialize, Debug, Default, Clone)]
+#[serde(deny_unknown_fields)]
+pub struct UpowerConfig {
+    /// Name of the power device to use.
+    #[serde(default = "UpowerConfig::default_device")]
+    pub device: Option<String>,
+}
+
+impl UpowerConfig {
+    fn default_device() -> Option<String> {
+        None
+    }
+}
+
+impl ConfigBlock for Upower {
+    type Config = UpowerConfig;
+
+    fn new(block_config: Self::Config, config: Config, send: Sender<Task>) -> Result<Self> {
+        let id: String = Uuid::new_v4().simple().to_string();
+        let id_copy = id.clone();
+        let dbus_conn = Connection::get_private(BusType::System)
+            .block_error("upower", "failed to establish D-Bus connection")?;
+        let battery = try!(match block_config.device {
+            Some(device) => Battery::from(&dbus_conn, device),
+            None => Battery::default(&dbus_conn),
+        });
+        let device_path = battery.get_device_path();
+
+        thread::spawn(move || {
+            let c = Connection::get_private(BusType::System).unwrap();
+            let rule = format!(
+                "type='signal',\
+                 path='{}',\
+                 interface='org.freedesktop.DBus.Properties',\
+                 member='PropertiesChanged'",
+                device_path);
+
+            c.add_match(&rule).unwrap();
+
+            loop {
+                let timeout = 100000;
+
+                for _event in c.iter(timeout) {
```

Comment:
Do you know how often dbus sends battery events? It's not a bad idea to have a minimum wait period to prevent event spam from triggering too many bar updates.

Reply:
Oh, that's a pretty substantial difference. I suppose that it's safe enough to dispense with a fallback, then.